### PR TITLE
Allow ContractInstances in @truffle/db without transactionHash information

### PR DESCRIPTION
**Pull request by gnidan**
This PR removes the non-nullability requirement for ContractInstance.creation.transactionHash to accommodate situations where it may be infeasible to obtain this information. As a result, ContractInstance.network may now represent any Network on which this instance is available at the end of that Network's historicBlock.

This means a couple things:
- `network` can no longer be an ID field, since this looser constraint means `network` is no longer unique; use `contract` and `creation` when computing ContractInstance IDs (in addition to the unchanged `address`).
- When doing `project.loadMigrate()`, if any input artifacts are missing a transactionHash, load the latest block as a Network resource and use that for any such ContractInstances.

### CODE REVIEW

File: packages/db/src/resources/types.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ export type Collections = {
   contractInstances: {
     resource: DataModel.ContractInstance;
     input: DataModel.ContractInstanceInput;
-    idFields: ["address", "network"];
+    idFields: ["contract", "address", "creation"];
```
Comment (fainashalts) [inline packages/db/src/resources/types.ts:55]: This is fine but we really should consider/plan a migration system post-haste. What if you hadn't realized the need for this until after we released DB into the wild?

**Review by fainashalts — APPROVED** [review]
This all LGTM but just wanna be sure I understand one mechanic: if we are missing the txHash and can't set up the Network resource like we normally would, we just fetch the latest that the DB has? But like, that could be wrong, couldn't it? Just wanna be sure we've considered some sort of way to convey that "this network wasn't computed the normal way" maybe? Maybe not, just a thought as I was reading this!

**Comment by gnidan** [comment]
> if we are missing the txHash [...] we just fetch the latest that the DB has?

In this situation, we fetch the latest from the chain and add that as its own Network resource (btw, I use "fetch" to mean "from the blockchain" and "query" to mean "from @truffle/db")

> Just wanna be sure we've considered some sort of way to convey that "this network wasn't computed the normal way" maybe?

I have a change in mind for this (define a ContractInstanceNetwork resource, to make ContractInstance <---> Network many-to-many), but that will have to wait. In the meantime, I think we just have to take `ContractInstance.network` as this semi-canonical record.

File: packages/db/src/resources/types.ts (hunk 1 above)
Comment (gnidan) [reply packages/db/src/resources/types.ts:55]: Yeah, I know :o 

See below... I already have another change that I foresee here, for after release :(

**Comment by gnidan** [comment]
Well, scary issues aside, gonna merge this! Thanks @fainashalts!

**Pull request merged**